### PR TITLE
Test: Refactor components/token-field tests to use enzyme

### PR DESCRIPTION
**Pull request by gziolo**
`components/token-field` tests got refactored to use `enzyme`.

This is also the slowest test suite. Execution time is on average 0.8s. I didn't make any progress performance wise. It looks like mouse and keyboard events are super slow, but I'm not an expert in this area.

### Testing
* `npm run test-client`
* `npm run test-client client/components/token-field/test/index.jsx -- -w`

### CODE REVIEW

File: client/components/token-field/test/index.jsx [hunk 1]
```diff
@@ -301,21 +292,22 @@ describe( 'TokenField', function() {
...
+			textInputNode.simulate( 'blur' );
+
+			// After blur, need to wait for TokenField#_blurTimeoutID
+			defer( function() {
 				testSavedState( false );
-				TestUtils.Simulate.focus( textInputNode );
+				textInputNode.simulate( 'focus' );
 				testSavedState( true );
 				done();
-			}, 10 );
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: `defer` is the [equivalent of `setTimeout( fn, 1 );`](https://github.com/lodash/lodash/blob/4.10.0/lodash.js#L9467), so not quite the same as was here previously. Not sure whether the 10 millisecond delay was chosen for any particular reason (cc @nylen), but would be happy to see it reduced if possible.

File: client/components/token-field/test/index.jsx [hunk 2]
```diff
@@ -32,89 +35,81 @@ const charCodes = {
...
-		textInputNode = tokenFieldNode.querySelector( '.token-field__input' );
-		TestUtils.Simulate.focus( textInputNode );
-	} );
-
-	afterEach( function() {
-		ReactDom.unmountComponentAtNode( reactContainer );
+		wrapper = mount( <TokenFieldWrapper /> );
+		tokenFieldNode = wrapper.ref( 'tokenField' );
+		textInputNode = tokenFieldNode.find( '.token-field__input' );
+		textInputNode.simulate( 'focus' )
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: Nitpick, but also to raise an aside: Should be a semi-colon here, and I'm at a loss for why the [`semi` rule](http://eslint.org/docs/rules/semi) isn't enabled in our ESLint configuration. At some point, we may have upgraded to an ESLint version where default rules were not specified, but it's one we should be enforcing ([related guideline](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#semicolons)).

File: client/components/token-field/test/index.jsx [hunk 3]
```diff
@@ -32,89 +35,81 @@ const charCodes = {
...
+		div.innerHTML = node.find( 'span' ).html();
+
+		return map( div.firstChild.childNodes, childNode => childNode.innerHTML );
 	}
 
 	function getSelectedSuggestion() {
 		var selectedSuggestions = getSuggestionsHTML( '.token-field__suggestion.is-selected' );
-		return selectedSuggestions[0] || null;
+
+		return selectedSuggestions[ 0 ] || null;
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: Would you back me up if I proposed removing [our exception from the spacing guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls)? :smile:

File: client/components/token-field/test/index.jsx [hunk 4]
```diff
@@ -32,89 +35,81 @@ const charCodes = {
...
+		return suggestionNodes.map( getSuggestionNodeHTML );
 	}
 
 	function getSuggestionNodeHTML( node ) {
-		var suggestionContent = node.childNodes[0];
-		if ( suggestionContent.nodeName === '#text' ) {
-			// This suggestion is not part of a partial match; just return
-			// the whole suggestion
-			return node.innerHTML;
+		if ( ! node.find( 'span' ).length ) {
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: Will this break when we upgrade to React 15.0 ? (Which I'd like to explore in the very near future)

Related: https://facebook.github.io/react/blog/2016/04/07/react-v15.html#no-more-extra-ltspangts

File: client/components/token-field/test/index.jsx [hunk 5]
```diff
@@ -233,63 +224,63 @@ describe( 'TokenField', function() {
 		it( 'should add a token when Tab pressed', function() {
 			setText( 'baz' );
 			sendKeyDown( keyCodes.tab );
-			expect( wrapper.state.tokens ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
-			expect( textInputNode.value ).to.equal( '' );
+			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
+			expect( textInputNode.props().value ).to.equal( '' );
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: FYI: You can get a single value via `textInputNode.prop( 'value' )`

File: client/components/token-field/test/index.jsx [hunk 6]
```diff
@@ -352,37 +344,40 @@ describe( 'TokenField', function() {
...
+			textInputNode.simulate( 'blur' );
 			setTimeout( function() {
 				// The click event comes from here the first time you click to
 				// activate the token field.
-				TestUtils.Simulate.click( wrapper.refs.tokenField.refs.tokensAndInput );
-				expect( wrapper.refs.tokenField.state.isActive ).to.equal( true );
+				tokenFieldNode.find( '.token-field__input-container' ).simulate( 'click' );
+				expect( tokenFieldNode.find( 'div' ).first().hasClass( 'is-active' ) ).to.equal( true );
 				done();
 			}, 50 ); // 50ms is a fast click
```
Comment (aduth) [inline client/components/token-field/test/index.jsx]: Do we still need the timeout if we're emulating a clock change?

**Comment by aduth** [comment]
A few questions, but otherwise confirm tests pass as expected. Marking as needing reply, but if you're comfortable with changes despite questions, feel free to merge. :+1:

File: client/components/token-field/test/index.jsx (hunk 1 above)
Comment (nylen) [inline client/components/token-field/test/index.jsx]: Should be fine to reduce this.

File: client/components/token-field/test/index.jsx (hunk 3 above)
Comment (nylen) [inline client/components/token-field/test/index.jsx]: I think the exception makes sense conceptually:  `arr[0]` is a simple, self-contained expression, whereas `arr[ arr.length - 1 ]` is more complicated and warrants a space to make it easier to parse.

However, we're not following our own standards here, so we might as well remove it:

- `ack '[a-z]\[ \d+\ ]' | wc -l`: 316
- `ack '[a-z]\[\d+\]' | wc -l`: 268

File: client/components/token-field/test/index.jsx (hunk 4 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: It's safe, we use `<span/>` explicitly: https://github.com/Automattic/wp-calypso/blob/master/client/components/token-field/suggestions-list.jsx#L100.

File: client/components/token-field/test/index.jsx (hunk 3 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: > Would you back me up if I proposed removing our exception from the spacing guidelines?

Sure, I guess this change was introduced by `Code reformat` feature I often use with PhpStorm/WebStorm. I don't have anything personally again similar exceptions, but it's really hard to teach such tools that there are exceptions :)

File: client/components/token-field/test/index.jsx (hunk 5 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: Good to know. Updating.

File: client/components/token-field/test/index.jsx (hunk 2 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: Fixed. Yes, there were no warnings from ESLint.

File: client/components/token-field/test/index.jsx (hunk 6 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: Also updated ;)

**Comment by gziolo** [comment]
I addressed most of the comments, refactored all `setTimeout` and `defer` occurrences replacing with `sinon.clock`. It's ready to review again.

File: client/components/token-field/test/index.jsx (hunk 3 above)
Comment (aduth) [inline client/components/token-field/test/index.jsx]: >I think the exception makes sense conceptually:  `arr[0]` is a simple, self-contained expression, whereas `arr[ arr.length - 1 ]` is more complicated and warrants a space to make it easier to parse.

On its own, I could see this being a valid argument. However, the mere existence of an exception to a rule puts an additional cognitive burden on the developer to recall and implement. And by your own findings, it seems the exception is frequently _not_ recalled, further worsening the scenario by the resulting inconsistencies.

Comment (aduth) [inline client/components/token-field/test/index.jsx]: Related: The exception also prevents us from using the ESLint built-in `space-in-brackets` rule ([see config](https://github.com/Automattic/wp-calypso/blob/5349c3bbf01cf42cd30a0fa918016d30c6e007aa/.eslintrc.js#L82-L83)).

**Comment by aduth** [comment]
Happy to see the `setTimeout`s gone. Looks good to me :+1:

File: client/components/token-field/test/index.jsx (hunk 3 above)
Comment (nylen) [inline client/components/token-field/test/index.jsx]: It seems pointless to argue for a rule when we already don't follow it, so feel free to ping me on a PR to get rid of the exception.

I don't think we should enforce `space-in-brackets` though - people will probably continue writing `[0]` out of habit.

**Comment by gziolo** [comment]
Thanks for review and merging :)

File: client/components/token-field/test/index.jsx (hunk 3 above)
Comment (gziolo) [inline client/components/token-field/test/index.jsx]: > I don't think we should enforce space-in-brackets though - people will probably continue writing [0] out of habit.

Unless we provide ESLint `fix` action for this flag, which should be possible :) Like I said, the best conventions are those that your IDE or another tool can control itself :)

**Pull request merged**